buffer.go

```diff
@@ -0,0 +1,139 @@
+package kinesis
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"html/template"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/gliderlabs/logspout/router"
+)
+
+// PutRecordsLimit is the maximum number of records allowed for a PutRecords request.
+var PutRecordsLimit = 500
+
+// RecordSizeLimit is the maximum allowed size per record.
+var RecordSizeLimit int = 1 * 1024 * 1024 // 1MB
+
+// PutRecordsSizeLimit is the maximum allowed size per PutRecords request.
+var PutRecordsSizeLimit int = 5 * 1024 * 1024 // 5MB
+
+type recordBuffer struct {
+	client   *kinesis.Kinesis
+	pKeyTmpl *template.Template
+	input    *kinesis.PutRecordsInput
+	count    int
+	byteSize int
+}
+
+func newRecordBuffer(client *kinesis.Kinesis, streamName string) (*recordBuffer, error) {
+	pKeyTmpl, err := pKeyTmpl()
+	if err != nil {
+		return nil, err
+	}
+
+	input := &kinesis.PutRecordsInput{
+		StreamName: aws.String(streamName),
+		Records:    make([]*kinesis.PutRecordsRequestEntry, 0),
+	}
+
+	return &recordBuffer{
+		client:   client,
+		pKeyTmpl: pKeyTmpl,
+		input:    input,
+	}, nil
+}
+
+func pKeyTmpl() (*template.Template, error) {
+	pKeyTmplString := os.Getenv("KINESIS_PARTITION_KEY_TEMPLATE")
+	if pKeyTmplString == "" {
+		return nil, errors.New("The partition key template is missing. Please set the KINESIS_PARTITION_KEY_TEMPLATE env variable")
+	}
+
+	pKeyTmpl, err := template.New("kinesisPartitionKey").Parse(pKeyTmplString)
+	if err != nil {
+		return nil, err
+	}
+
+	return pKeyTmpl, nil
+}
+
+func (r *recordBuffer) Add(m *router.Message) error {
+	data := m.Data
+	dataLen := len(data)
+
+	// This record is too large, we can't submit it to kinesis.
+	if dataLen > RecordSizeLimit {
+		return errors.New(fmt.Sprintf("recordBuffer.Add: log data byte size (%d) is over the limit.", dataLen))
```

Comment:
Also - what happens to these errors?  Do they get bubbled up to us any way?  I'm worried we'll miss these.

We could, in theory, break the message up over multiple entries, right?  Not sure this is the right thing to do - just something to think about.